pulsar-client-go/examples/consumer-listener/consumer-listener.go

```diff
@@ -0,0 +1,55 @@
+//
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+//
+
+package main
+
+import (
+	"../../pulsar"
+	"fmt"
+	"log"
+)
+
+func main() {
+	client := pulsar.NewClient().
+		ServiceUrl("pulsar://localhost:6650").Build()
+
+	channel := make(chan pulsar.ConsumerMessage)
+
+	consumer, err := client.NewConsumer().
```

Comment:
The builder pattern is not very idiomatic in Go. A struct whose values can be set is simpler and generally preferred. I am less familiar with how that would impact the cgo requirement (I have not written any cgo myself). But I can imagine once `Subscribe` is called which sends the network call, any underlying cgo logic can be performed along with setting defaults.

In other words, this pattern would be ideal (also note removing the field name redundancy of `Subscription`.. names are heavily driven by context in which they are used. If something sounds redundant it should be reduced).

```go
consumer := &Consumer{
  Topic: "my-topic",
  Name: "my-subscription",
  Type: pulsar.Shared,
  Channel: channel,
}

if err := consumer.Subscribe(); err != nil {
  log.Fatal(err)
}

// ...
```

Reply:
Changing to a struct-based pattern over a builder obviously applies to the other APIs in this PR as well..

Reply:
Yes, I was trying to use the struct approach for config but I could not find a good way to deal with integer defaults.

Example: producer configuration has a "send timeout" int option, for which a sensible default is 30 seconds. The problem is that the struct default of 0 is a valid value too, so I couldn't find a way to differentiate a legit 0 from an uninitiated option

Reply:
Hm, yes default zero values can be a nuisance. There are a couple patterns for this case, but they are escaping me at the moment. One more general alternative is to stick with the `New*` functions to initialize a client, consumer, producer, and reader value with the defaults preset on the struct which then allows the user to override the options they want. The `New` call could take the required options.

```go
// Constructor with required options.
c := NewConsumer(topic)
// Override default
c.Timeout = 10*time.Second

// Rest..
if err := c.Subscribe(); err != nil {
}
```

So in this set, trading the builder pattern for a simple struct that is returned that can be mutated before making the connection.